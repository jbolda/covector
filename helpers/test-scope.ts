import { type Operation, run } from "effection";

export interface TestScope {
  addSetup(op: () => Operation<void>): void;
  runTest(op: () => Operation<void>): Promise<void>;
}

export function createTestScope(): TestScope {
  let setup: (() => Operation<void>)[] = [];
  return {
    addSetup(op) {
      setup.push(op);
    },
    runTest(op) {
      return run(function* () {
        for (let step of setup) {
          yield step();
        }
        yield op();
      });
    },
  };
}

import * as vitest from "vitest";

let scope: TestScope | undefined;

function describeWithScope(
  name: string | Function,
  factory?: vitest.SuiteFactory<{}>
): vitest.SuiteCollector<{}> {
  return vitest.describe(name, () => {
    vitest.beforeEach(() => {
      if (!scope) {
        scope = createTestScope();
      }
    });
    if (factory && typeof factory === "function") {
      (<Function>factory)();
    }
  });
}

describeWithScope.only = vitest.describe.only;
describeWithScope.skip = vitest.describe.skip;
describeWithScope.skipIf = vitest.describe.skipIf;

export const describe = <typeof vitest.describe>(<unknown>describeWithScope);

export function beforeEach(op: () => Operation<void>): void {
  vitest.beforeEach(() => scope!.addSetup(op));
}

export function it(
  desc: string,
  op?: () => Operation<void>,
  timeout?: number
): void {
  if (op) {
    return vitest.it(desc, async () => scope?.runTest(op), timeout);
  } else {
    return vitest.it.skip(desc, () => {});
  }
}

it.only = function only(
  desc: string,
  op?: () => Operation<void>,
  timeout?: number
): void {
  if (op) {
    return vitest.it.only(desc, async () => scope!.runTest(op), timeout);
  } else {
    return vitest.it.skip(desc, () => {});
  }
};

export function captureError<T>(op: Operation<T>): Operation<T | Error> {
  return function* () {
    try {
      yield op;
    } catch (error) {
      return error;
    }
    throw new Error("expected operation to throw an error, but it did not!");
  };
}
