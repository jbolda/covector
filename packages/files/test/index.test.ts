import { describe, it } from "../../../helpers/test-scope.ts";
import { expect } from "vitest";
import pino from "pino";
import * as pinoTest from "pino-test";
import fixtures from "fixturez";

import {
  readPkgFile,
  readAllPkgFiles,
  readPreFile,
  configFile,
  changeFiles,
  loadChangeFiles,
  changeFilesRemove,
} from "../src";

const f = fixtures(__dirname);

describe("general file test", () => {
  it("parses general file", function* () {
    const generalFolder = f.copy("pkg.general-file");
    const generalFile = yield readPkgFile({
      file: "VERSION",
      cwd: generalFolder,
      nickname: "general-package",
    });
    expect(generalFile.name).toBe("general-package");
    expect(generalFile.version).toBe("6.1.0");
  });

  it("parses config", function* () {
    const configFolder = f.copy("config.simple");
    const configArray = yield configFile({ cwd: configFolder });
    expect((configArray as any).gitSiteUrl).toBe(
      "https://github.com/jbolda/covector"
    );
  });

  it("reads all package files in config", function* () {
    const configFolder = f.copy("integration.js-with-subcommands");
    const config = yield configFile({ cwd: configFolder });
    const allPackages = yield readAllPkgFiles({ config, cwd: configFolder });
    expect(allPackages["package-one"].name).toBe("package-one");
    expect(allPackages["package-one"].version).toBe("2.3.1");

    expect(allPackages["package-two"].name).toBe("package-two");
    expect(allPackages["package-two"].version).toBe("1.9.0");
  });

  describe("parses pre", () => {
    it("parses pre without changes", function* () {
      const preFolder = f.copy("pre.without-changes");
      const preFile = yield readPreFile({ cwd: preFolder });
      expect(preFile?.tag).toBe("beta");
      expect(preFile?.changes.length).toBe(0);
    });

    it("parses pre with changes", function* () {
      const preFolder = f.copy("pre.with-changes");
      const preFile = yield readPreFile({ cwd: preFolder });
      expect(preFile?.tag).toBe("beta");
      expect(preFile?.changes.length).toBe(3);
      expect(preFile?.changes[1]).toBe("chocolate-pudding.md");
    });

    it("returns cleanly without pre", function* () {
      const preFolder = f.copy("pkg.js-basic");
      const preFile = yield readPreFile({ cwd: preFolder });
      expect(preFile).toBe(null);
    });
  });

  it("globs changes", function* () {
    const changesFolder = f.copy("changes.multiple-changes");
    const changesPaths = yield changeFiles({ cwd: changesFolder });
    const changesFiles = yield loadChangeFiles({
      cwd: changesFolder,
      paths: changesPaths,
    });
    expect(changesFiles).toMatchSnapshot();
  });

  it("ignores readme", function* () {
    const changesFolder = f.copy("changes.no-changes-with-readme");
    const changesArray = yield changeFiles({ cwd: changesFolder });
    expect(changesArray).toMatchSnapshot();
  });

  it("deletes files", function* () {
    const stream = pinoTest.sink();
    const logger = pino(stream);
    const changesFolder = f.copy("integration.general-file");
    const changeFilesToDelete = [
      "./.changes/first-change.md",
      "./.changes/second-change.md",
    ];

    const filesRemoved = yield changeFilesRemove({
      logger,
      cwd: changesFolder,
      paths: changeFilesToDelete,
    });

    expect(filesRemoved).toEqual(changeFilesToDelete);
  });
});
