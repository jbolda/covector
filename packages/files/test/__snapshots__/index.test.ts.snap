// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`general file test > globs changes 1`] = `
[
  {
    "content": "---
"package-a": minor
---

Summary about the changes in pkg-a
",
    "extname": ".md",
    "filename": "first-change",
    "path": ".changes/first-change.md",
  },
  {
    "content": "---
"package-b": patch
---

Summary about the changes in pkg-b
",
    "extname": ".md",
    "filename": "second-change",
    "path": ".changes/second-change.md",
  },
]
`;

exports[`general file test > ignores readme 1`] = `[]`;
